pilot/pkg/model/push_context.go

```diff
@@ -412,10 +418,30 @@ func (ps *PushContext) VirtualServices(proxy *Proxy, gateways map[string]bool) [
 
 // DestinationRule returns a destination rule for a service name in a given domain.
 func (ps *PushContext) DestinationRule(proxy *Proxy, hostname Hostname) *Config {
-	// TODO: use the proxy namespace to return only public destination rules if in different namespace
-	if c, ok := MostSpecificHostMatch(hostname, ps.destinationRuleHosts); ok {
-		return ps.destinationRuleByHosts[c].config
+	if proxy == nil {
+		// TODO: this may be not correct, but we can not get ConfigNamespace,
+		// so match public rule first
+		if host, ok := MostSpecificHostMatch(hostname, ps.publicDestRuleHosts); ok {
+			return ps.publicDestRuleByHost[host].config
+		}
+
+		for ns, privateDestHosts := range ps.privateDestRuleHostsByNamespace {
+			if host, ok := MostSpecificHostMatch(hostname, privateDestHosts); ok {
+				return ps.privateDestRuleByHostByNamespace[ns][host].config
+			}
```

Comment:
I think you can skip the proxy==nil check (if all callsites are providing the proxy, then you can simply do if proxy==nil, return nil)